index.js

```diff
@@ -231,4 +231,55 @@ transform.resetTypes = function resetTypes() {
   registerTypes = {};
 };
 
-module.exports = transform;
+function addPermissions(permissions, key, type) {
+  return permissions[key] && permissions[key][type] !== undefined ? permissions[key][type] : permissions[type]
+}
+
+function getFormOptions(properties, fieldsData = {}, permissions, newOptions = {}, first = false) {
+  Object.entries(properties).forEach((data) => {
+      const [key, value] = data
+      if (first) {
+          newOptions[key] = {
+              fields: [],
+              order: fieldsData[key] ? fieldsData[key].order : []
+          }
+      } else {
+          // If field doesn't exist, create it
+          if (!fieldsData[key]) {
+              fieldsData[key] = {}
+          }
+          if (!fieldsData[key]["ui:component"] || fieldsData[key]["ui:component"].includes('auto-fill')) {
+              fieldsData[key]["ui:component"] = 'string'
+          }
+          newOptions[key] = {
+              ...fieldsData[key]
+          }
+      }
+      // Add permissions
+      newOptions[key] = {
+          ...newOptions[key],
+          editable: addPermissions(permissions, key, 'editable'),
+          viewable: addPermissions(permissions, key, 'viewable')
+      }
+      if (value.properties || (value.items && value.items.properties)) {
+          const prop = value.items && value.items.properties ? value.items.properties : value.properties
+          const opt = newOptions[key].fields ? newOptions[key].fields : newOptions[key]
+          let fields = {}
+          if (fieldsData[key]) {
+              fields = fieldsData[key].fields ? fieldsData[key].fields : fieldsData[key]
+          }
+          const newPermissions = {
+              ...permissions[key],
+              editable: addPermissions(permissions, key, 'editable'),
+              viewable: addPermissions(permissions, key, 'viewable')
+          }
+          getFormOptions(prop, fields, newPermissions, opt)
```

Comment:
Don't we need to add the return value to `newOptions`?

Reply:
From a glance, it seems like this should be something more like:
```
newOptions[key][prop] = getFormOptions(prop, fields, newPermissions, opt)
```
not just:
```
return getFormOptions(prop, fields, newPermissions, opt)
```
but ignore me if it's wrong. Once you add some tests, it will be easier to discuss any issues with the behavior of this function.